### PR TITLE
Enable Synchronous Child Datastore Creation

### PR DESCRIPTION
**Pull request by anthony-murphy**
## Overview

This feature introduces a new pattern for creating datastores synchronously within the Fluid Framework. It allows for the synchronous creation of a child datastore from an existing datastore, provided that the child datastore is available synchronously via the existing datastore's registry and that the child's factory supports synchronous creation. This method also ensures strong typing for the consumer.

In this context, "child" refers specifically to the organization of factories and registries, not to any hierarchical or hosting relationship between datastores. The parent datastore does not control the runtime behaviors of the child datastore beyond its creation.

The synchronous creation of child datastores enhances the flexibility of datastore management within the Fluid Framework. It ensures type safety and provides a different way to manage datastores within a container. However, it is important to consider the overhead associated with datastores, as they are stored, summarized, garbage collected, loaded, and referenced independently. This overhead should be justified by the scenario's requirements.

Datastores offer increased capabilities, such as the ability to reference them via handles, allowing multiple references to exist and enabling those references to be moved, swapped, or changed. Additionally, datastores are garbage collected after becoming unreferenced, which can simplify final cleanup across clients. This is in contrast to subdirectories in a shared directory, which do not have native capabilities for referencing or garbage collection but are very low overhead to create.

Synchronous creation relies on both the factory and the datastore to support it. This means that asynchronous operations, such as resolving handles, some browser API calls, consensus-based operations, or other asynchronous tasks, cannot be performed during the creation flow. Therefore, synchronous child datastore creation is best limited to scenarios where the existing asynchronous process cannot be used, such as when a new datastore must be created in direct response to synchronous user input.

## Key Benefits

- **Synchronous Creation**: Allows for the immediate creation of child datastores without waiting for asynchronous operations.
- **Strong Typing**: Ensures type safety and better developer experience by leveraging TypeScript's type system.

### Example 1: Creating a Child Datastore

In this example, we demonstrate how to support creating a child datastore synchronously from a parent datastore.

```typescript
/**
 * This is the parent DataObject, which is also a datastore. It has a
 * synchronous method to create child datastores, which could be called
 * in response to synchronous user input, like a key press.
 */
class ParentDataObject extends DataObject {
	get ParentDataObject() {
		return this;
	}
	protected override async initializingFirstTime(): Promise<void> {
		// create synchronously during initialization
		this.createChild("parentCreation");
	}

	createChild(name: string): ChildDataStore {
		assert(
			this.context.createChildDataStore !== undefined,
			"this.context.createChildDataStore",
		);
		// creates a detached context with a factory who's package path is the same
		// as the current datastore, but with another copy of its own type.
		const { entrypoint } = this.context.createChildDataStore(
			ChildDataStoreFactory.instance,
		);
		const dir = this.root.createSubDirectory("children");
		dir.set(name, entrypoint.handle);
		entrypoint.setProperty("childValue", name);

		return entrypoint;
	}

	getChild(name: string): IFluidHandle<ChildDataStore> | undefined {
		const dir = this.root.getSubDirectory("children");
		return dir?.get<IFluidHandle<ChildDataStore>>(name);
	}
}
```

For a complete example see the follow test:
https://github.com/microsoft/FluidFramework/blob/main/packages/test/local-server-tests/src/test/synchronousDataStoreCreation.spec.ts

### CODE REVIEW

**Comment by anthony-murphy** [comment]
This change brings back #22962 which was reverted for a minor incompatibility issue, which fully resolved in this version by splitting the get method on IFluidDataStoreRegistery into two, get and getSync.

File: .changeset/solid-keys-agree.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	getChild(name: string): IFluidHandle<ChildDataStore> | undefined {
+		const dir = this.root.getSubDirectory("children");
+		return dir?.get<IFluidHandle<ChildDataStore>>(name);
+	}
+}
+```
+
+For a complete example see the follow test:
```
Comment (jzaffiro) [inline .changeset/solid-keys-agree.md:73]:
```suggestion
For a complete example see the following test:
```

**Review by jzaffiro — APPROVED** [review]

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts [hunk 2]
```diff
@@ -593,6 +596,28 @@ export interface IFluidDataStoreContext extends IFluidParentContext {
 	 * and its children with the GC details from the previous summary.
 	 */
 	getBaseGCDetails(): Promise<IGarbageCollectionDetailsBase>;
+
+	/**
+	 * Synchronously creates a detached child data store.
+	 *
+	 * The `createChildDataStoreSync` method allows for the synchronous creation of a detached child data store. This is particularly
```
Comment (Josmithr) [inline packages/runtime/runtime-definitions/src/dataStoreContext.ts:603]: Nit: this references `createChildDataStoreSync`, but the method is named `createChildDataStore`

File: packages/runtime/runtime-definitions/src/dataStoreFactory.ts [hunk 3]
```diff
@@ -21,24 +21,63 @@ export interface IProvideFluidDataStoreFactory {
...
+	 * Uniquely identifies the type of data store created by this factory.
 	 */
 	type: string;
 
 	/**
-	 * Generates runtime for the data store from the data store context. Once created should be bound to the context.
-	 * @param context - Context for the data store.
-	 * @param existing - If instantiating from an existing file.
+	 * Asynchronously generates the runtime for the data store from the given context.
+	 * Once created, the data store should be bound to the context.
```
Comment (Josmithr) [inline packages/runtime/runtime-definitions/src/dataStoreFactory.ts:48]: Nit
```suggestion
	 * @remarks
	 * Once created, the data store should be bound to the context.
```

File: packages/runtime/runtime-definitions/src/dataStoreFactory.ts [hunk 4]
```diff
@@ -21,24 +21,63 @@ export interface IProvideFluidDataStoreFactory {
...
 	 */
 	instantiateDataStore(
 		context: IFluidDataStoreContext,
 		existing: boolean,
 	): Promise<IFluidDataStoreChannel>;
+
+	/**
+	 * Synchronously creates a new runtime for a new data store from the provided context.
+	 *
+	 * This method enables a synchronous creation path. Specifically, if this factory is registered
```
Comment (Josmithr) [inline packages/runtime/runtime-definitions/src/dataStoreFactory.ts:67]: Nit
```suggestion
	 * @remarks
	 * This method enables a synchronous creation path. Specifically, if this factory is registered
```

**Review by Josmithr — APPROVED** [review]

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:603]:
```suggestion
	 * The `createChildDataStore` method allows for the synchronous creation of a detached child data store. This is particularly
```

**Pull request merged**